lib/net/ldap/connection.rb

```diff
@@ -8,56 +8,49 @@ class Net::LDAP::Connection #:nodoc:
 
   def initialize(server)
     @instrumentation_service = server[:instrumentation_service]
-    server[:hosts] = [[server[:host], server[:port]]] if server[:hosts].nil?
 
     if server[:socket]
-      prepare_socket(server)
+      prepare_socket(server[:socket], server[:encryption])
     else
-      open_connection(server)
+      hosts = server[:hosts]
+      hosts = [[server[:host], server[:port]]] if hosts.nil?
+      open_connection(hosts, server[:encryption])
```

Comment:
I should've suggested this in your previous PR, but what do you think of this idea: Instead of having `open_connection` worry about handling multiple hosts and then checking whether it did one host, or multiple, we can iterate through `server[:hosts]` here. The advantage is that `open_connection` would only be responsible for what it's method name suggests: opening a single connection. You can then collect the errors here and handle it in a generic way without the code smell we're discussing below. It'd also minimize remove the changes from `open_connection`, which I think make the method less readable.